src/index.ts

```diff
@@ -0,0 +1,123 @@
+import * as Promise from 'bluebird'
+import * as _ from 'lodash'
+import * as tar from 'tar-stream'
+
+import Bundle from './bundle'
+import { FileInfo } from './fileInfo'
+import { Resolver } from './resolver'
+import * as Utils from './utils'
+
+// Import some default resolvers
+import DockerfileResolver from './resolvers/dockerfile'
+import DockerfileTemplateResolver from './resolvers/dockerfileTemplate'
+import ArchDockerfileResolver from './resolvers/archDockerfile'
+
+// re-export
+export { ArchDockerfileResolver,
+         Bundle,
+				 DockerfileResolver,
+         DockerfileTemplateResolver,
+         FileInfo,
+         Resolver }
+
+export interface ResolvedBundle {
+	projectType: string,
+	tarStream: tar.Pack
+}
+
+export function resolveBundle(bundle: Bundle, resolvers: Resolver[]): Promise<ResolvedBundle> {
+	return new Promise<ResolvedBundle>((resolve: (resolved: ResolvedBundle) => void,
+                                		  reject: (err: Error) => void) => {
+		const extract = tar.extract()
+		const pack = tar.pack()
+
+		extract.on('entry', (header: tar.TarHeader, stream: NodeJS.ReadableStream, next: () => void) => {
+			// Read the contents into a buffer
+			let buf: string = ''
+
+			stream.on('data', (data: string) => {
+				buf += data
+			})
+
+			stream.on('end', () => {
+				// Send the file along to the next tar stream regardless
+				pack.entry(header, buf)
+
+				// create a FileInfo from the header
+				const info: FileInfo = {
+					name: Utils.normalizeTarEntry(header.name),
+					size: header.size,
+					contents: buf
+				}
+
+				// Now provide the resolvers with the information and file
+				resolvers.map((resolver) => {
+					resolver.entry(info)
+				})
+			})
+
+			next()
+		})
+
+		extract.on('finish', () => {
+			let maybeResolver: Resolver | undefined = undefined
+			// Check to see if any of the resolvers are happy, in `priority` order
+			_(_.orderBy(resolvers, (val: Resolver) => val.priority, ['desc']))
+			.each((resolver: Resolver) => {
```

Comment:
Don't need to manage the `maybeResolver`/`return` part yourself, and you can chain lodash methods too, to make this all nicer. You should be able to replace this whole each block with something like:

```javascript
let maybeResolver = _(resolvers)
    .orderBy((val: resolver) => val.priority, ['desc'])
    .find((resolver: Resolver) => resolver.isSatisfied(bundle))
```